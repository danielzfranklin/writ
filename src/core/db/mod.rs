pub mod author;
pub mod blob;
pub mod cache;
pub mod commit;
pub mod object;
pub mod tree;

pub use author::Author;
pub use blob::Blob;
pub use commit::Commit;
pub use object::{Object, ObjectBuilder, Oid, UntypedOid};
pub use tree::Tree;

use bstr::{BString, ByteSlice};
use flate2::{read::ZlibDecoder, write::ZlibEncoder, Compression};
use tempfile::NamedTempFile;

use std::{
    collections::BTreeMap,
    fs::{self, File},
    io::{self, BufRead, BufReader, BufWriter, ErrorKind, Read, Write},
    num::ParseIntError,
    path::PathBuf,
};

use self::cache::Cache;
use crate::core::WsPath;

/// Note: Cloning doesn't keep the cache
#[derive(Debug)]
pub struct Db {
    path: PathBuf,
    cache: Cache,
}

impl Db {
    pub fn new<P: Into<PathBuf>>(git_dir: P) -> Self {
        Self {
            path: git_dir.into().join("objects"),
            cache: Cache::new(),
        }
    }

    pub fn load<O: Object>(&mut self, oid: Oid<O>) -> Result<O, LoadError<O>> {
        if let Some(cached) = self.cache.get(&oid) {
            return Ok(cached.clone());
        }

        let (len, bytes) = self.load_bytes(O::TYPE, &oid)?;
        let object = O::deserialize(oid, len, bytes).map_err(|e| LoadError::Deserialize(oid, e))?;
        self.cache.insert(oid, object.clone());

        Ok(object)
    }

    pub fn load_tree_file(
        &mut self,
        mut tree: Oid<Tree>,
        path: &WsPath,
    ) -> Result<Option<tree::FileNode>, LoadError<Tree>> {
        for parent in path.parents() {
            if let Some(tree::Node::Tree { oid: next_tree, .. }) =
                self.load(tree)?.direct_child(parent.file_name())
            {
                tree = *next_tree;
            } else {
                return Ok(None);
            }
        }

        if let Some(tree::Node::File(file)) = self.load(tree)?.direct_child(path.file_name()) {
            Ok(Some(file.clone()))
        } else {
            Ok(None)
        }
    }

    pub fn load_tree_files(
        &mut self,
        root: &WsPath,
        tree: Oid<Tree>,
    ) -> Result<BTreeMap<WsPath, tree::FileNode>, LoadError<Tree>> {
        let mut files = BTreeMap::new();
        self.load_all_tree_files_into(&mut files, root, tree)?;
        Ok(files)
    }

    fn load_all_tree_files_into(
        &mut self,
        files: &mut BTreeMap<WsPath, tree::FileNode>,
        root: &WsPath,
        tree: Oid<Tree>,
    ) -> Result<(), LoadError<Tree>> {
        let tree = self.load(tree)?;
        for child in tree.direct_children() {
            match child {
                tree::Node::File(file) => {
                    let path = root.join_bytes(child.name());
                    files.insert(path, file.clone());
                }
                tree::Node::Tree { oid, .. } => {
                    let next_root = root.join_bytes(child.name());
                    self.load_all_tree_files_into(files, &next_root, *oid)?;
                }
            }
        }
        Ok(())
    }

    /// Doesn't cache
    fn load_bytes<O: Object>(
        &self,
        expected_type: &[u8],
        oid: &Oid<O>,
    ) -> Result<(usize, impl BufRead), LoadBytesError<O>> {
        let path = self.oid_path(&oid);
        let file = match File::open(&path) {
            Ok(file) => Ok(file),
            Err(err) if err.kind() == io::ErrorKind::NotFound => {
                Err(LoadBytesError::NotFound(*oid))
            }
            Err(err) => Err(LoadBytesError::Open(*oid, err)),
        }?;

        let mut bytes = BufReader::new(ZlibDecoder::new(file));

        let mut o_type = vec![0; expected_type.len() + 1];
        bytes
            .read_exact(&mut o_type)
            .map_err(|e| LoadBytesError::ReadPrefix(*oid, e))?;
        let sep = o_type.pop().unwrap();
        if sep != b' ' {
            return Err(LoadBytesError::Corrupt(*oid));
        }
        if o_type != expected_type {
            return Err(LoadBytesError::WrongType {
                oid: *oid,
                expected: expected_type.into(),
                actual: o_type.into(),
            });
        }

        let mut len = BString::from(Vec::new());
        bytes
            .read_until(b'\0', &mut len)
            .map_err(|e| LoadBytesError::ReadPrefix(*oid, e))?;
        len.pop().unwrap();
        let len = len
            .to_str()
            .map_err(|e| LoadBytesError::ParseLenToBytes(*oid, e))?;
        let len: usize = len
            .parse()
            .map_err(|e| LoadBytesError::ParseLenToInt(*oid, e))?;

        Ok((len, bytes))
    }

    /// Doesn't cache
    pub fn store_bytes<OB: ObjectBuilder>(&self, content: &[u8]) -> StoreResult<OB::Object> {
        fn helper<O: Object>(db: &Db, oid: &Oid<O>, bytes: &[u8]) -> io::Result<()> {
            let path = db.oid_path(oid);

            if path.exists() {
                return Ok(());
            }

            let mut temp = NamedTempFile::new()?;

            {
                let mut writer = BufWriter::new(&mut temp);
                let mut writer = ZlibEncoder::new(&mut writer, Compression::default());
                writer.write_all(bytes)?;
            }

            // We use a temp file to get an atomic write
            temp.flush()?;

            match fs::rename(temp.path(), &path) {
                Err(err) if err.kind() == ErrorKind::NotFound => {
                    fs::create_dir(&path.parent().expect("has parent"))?;
                    fs::rename(temp.path(), &path)?;
                }
                Err(err) => return Err(err),
                Ok(()) => (),
            }

            Ok(())
        }

        let o_type = OB::Object::TYPE;

        let mut bytes = Self::serialized_prefix(o_type, content);
        bytes.extend_from_slice(content);
        let oid = Oid::for_serialized_bytes(&bytes);

        helper::<OB::Object>(self, &oid, &bytes).map_err(|e| StoreError(oid, e))?;

        Ok(oid)
    }

    fn serialized_prefix(o_type: &[u8], serialized: &[u8]) -> Vec<u8> {
        let size = serialized.len().to_string();

        let mut ser = Vec::with_capacity(o_type.len() + 1 + size.len() + 1);

        ser.extend(o_type);
        ser.push(b' ');
        ser.extend(size.as_bytes());
        ser.push(b'\0');

        ser
    }

    fn oid_path<O: Object>(&self, oid: &Oid<O>) -> PathBuf {
        let oid = oid.to_hex();
        let dir = self.path.join(&oid[0..2]);
        let name = &oid[2..];
        dir.join(name)
    }
}

impl Clone for Db {
    fn clone(&self) -> Self {
        Self::new(&self.path)
    }
}

pub type StoreResult<O> = Result<Oid<O>, StoreError<O>>;

#[derive(Debug, thiserror::Error, displaydoc::Display)]
/// Failed to store {0:?}
pub struct StoreError<O: Object>(Oid<O>, #[source] io::Error);

#[derive(Debug, thiserror::Error, displaydoc::Display)]
pub enum LoadError<O: Object + 'static> {
    /// Failed to load bytes of object {0:?}
    LoadBytes(#[from] LoadBytesError<O>),
    /// Failed to deserialize {0:?}
    Deserialize(Oid<O>, #[source] O::DeserializeError),
}

#[derive(Debug, thiserror::Error, displaydoc::Display)]
pub enum LoadBytesError<O: Object + 'static> {
    /// {0:?} not found in database
    NotFound(Oid<O>),
    /// Failed to open the file for {0:?} in the database
    Open(Oid<O>, #[source] io::Error),
    /// Failed to read the prefix from the file for {0:?} in the database
    ReadPrefix(Oid<O>, #[source] io::Error),
    /// Database entry for {0:?} is corrupt
    Corrupt(Oid<O>),
    /// Expected oid {oid:?} to have type {expected}, got {actual}
    WrongType {
        oid: Oid<O>,
        expected: BString,
        actual: BString,
    },
    /// Failed to parse bytes of length of {0:?} as utf-8
    ParseLenToBytes(Oid<O>, #[source] bstr::Utf8Error),
    /// Failed to parse length of {0:?}
    ParseLenToInt(Oid<O>, #[source] ParseIntError),
}
