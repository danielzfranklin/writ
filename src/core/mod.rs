pub mod db;
pub mod index;
pub mod locked_file;
pub mod refs;
pub mod repo;
pub mod stat;
pub mod status;
pub mod with_digest;
pub mod ws;

pub use db::{Db, Object, ObjectBuilder, Oid};
pub use index::{Index, IndexMut};
pub use locked_file::LockedFile;
pub use refs::Refs;
pub use repo::Repo;
pub use stat::Stat;
pub use status::{FileStatus, Status};
pub use with_digest::WithDigest;
pub use ws::Workspace;
pub use ws::WsPath;
