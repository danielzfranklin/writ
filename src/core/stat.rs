use std::{
    convert::TryInto,
    fs,
    os::linux::fs::MetadataExt,
    time::{Duration, SystemTime},
};
use tracing::warn;

use bstr::{BStr, ByteSlice};

#[derive(Debug, Clone, Copy, Eq, PartialEq)]
pub struct Stat {
    pub ctime: SystemTime,
    pub mtime: SystemTime,
    pub dev: u32,
    pub ino: u32,
    pub mode: Mode,
    pub uid: u32,
    pub gid: u32,
    pub size: u32,
}

#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub enum Mode {
    Regular,
    Executable,
}

impl Stat {
    #[allow(
        clippy::cast_possible_truncation,
        clippy::cast_sign_loss,
        clippy::similar_names
    )]
    pub fn from(meta: &fs::Metadata) -> Self {
        Self {
            ctime: Self::systemtime_from_epoch(meta.st_ctime() as u32, meta.st_ctime_nsec() as u32),
            mtime: Self::systemtime_from_epoch(meta.st_mtime() as u32, meta.st_mtime_nsec() as u32),
            dev: meta.st_dev() as u32,
            ino: meta.st_ino() as u32,
            mode: Mode::from_u32(meta.st_mode()),
            uid: meta.st_uid(),
            gid: meta.st_gid(),
            size: meta.st_size() as u32,
        }
    }

    pub fn zeroed() -> Self {
        Self {
            ctime: SystemTime::UNIX_EPOCH,
            mtime: SystemTime::UNIX_EPOCH,
            dev: 0,
            ino: 0,
            mode: Mode::Regular,
            uid: 0,
            gid: 0,
            size: 0,
        }
    }

    pub fn ctime_epoch(&self) -> (u32, u32) {
        Self::systemtime_to_epoch(self.ctime)
    }

    pub fn mtime_epoch(&self) -> (u32, u32) {
        Self::systemtime_to_epoch(self.mtime)
    }

    fn systemtime_to_epoch(time: SystemTime) -> (u32, u32) {
        let dur = time
            .duration_since(SystemTime::UNIX_EPOCH)
            .expect("Not before epoch");

        let secs: u32 = dur.as_secs().try_into().expect("Time overflowed");

        (secs, dur.subsec_nanos())
    }

    pub(crate) fn systemtime_from_epoch(secs: u32, nanos: u32) -> SystemTime {
        SystemTime::UNIX_EPOCH + Duration::new(u64::from(secs), nanos)
    }
}

impl Mode {
    const EXECUTABLE: u32 = 0o10_07_55;
    const REGULAR: u32 = 0o10_06_44;

    const REGULAR_S: &'static [u8] = b"100644";
    const EXECUTABLE_S: &'static [u8] = b"100755";

    pub fn as_base8(self) -> &'static BStr {
        match self {
            Self::Regular => Self::REGULAR_S.as_bstr(),
            Self::Executable => Self::EXECUTABLE_S.as_bstr(),
        }
    }

    pub fn as_u32(self) -> u32 {
        match self {
            Self::Regular => Self::REGULAR,
            Self::Executable => Self::EXECUTABLE,
        }
    }

    pub fn from_u32(val: u32) -> Self {
        let is_executable = val & 0o111 != 0;
        if is_executable {
            Self::Executable
        } else {
            Self::Regular
        }
    }

    /// Unrecognized modes are considered [`Self::Regular`]
    pub fn from_base8(bytes: &BStr) -> Self {
        match bytes.as_bytes() {
            Self::REGULAR_S => Self::Regular,
            Self::EXECUTABLE_S => Self::Executable,
            _ => {
                warn!("Assuming unrecognized mode {} to be regular", bytes);
                Self::Regular
            }
        }
    }
}
