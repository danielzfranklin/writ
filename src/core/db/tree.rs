use std::{
    collections::BTreeMap,
    io::{self, BufRead},
};

use bstr::{BStr, BString, ByteSlice};

use crate::core::{db, stat, Db, Object, Oid, WsPath};

use super::{object::OID_SIZE, Blob, ObjectBuilder, UntypedOid};

#[derive(Debug, Clone, Eq, PartialEq)]
pub struct Tree {
    oid: Oid<Tree>,
    nodes: BTreeMap<BString, Node>,
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub enum Node {
    File(FileNode),
    Tree { name: BString, oid: Oid<Tree> },
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub struct FileNode {
    pub oid: Oid<Blob>,
    pub name: BString,
    pub mode: stat::Mode,
}

impl Tree {
    const MODE: &'static [u8] = b"40000";

    pub fn direct_children(&self) -> impl Iterator<Item = &Node> {
        self.nodes.values()
    }

    pub fn direct_child(&self, name: &BStr) -> Option<&Node> {
        self.nodes.get(name)
    }
}

impl Object for Tree {
    const TYPE: &'static [u8] = b"tree";

    type Builder = Builder;
    type DeserializeError = DeserializeError;

    fn oid(&self) -> Oid<Self> {
        self.oid
    }

    fn deserialize(
        oid: Oid<Self>,
        _len: usize,
        mut data: impl std::io::BufRead,
    ) -> Result<Self, Self::DeserializeError> {
        let mut nodes = BTreeMap::new();

        while let Some(node) = Node::deserialize(&mut data)? {
            nodes.insert(node.name().to_owned(), node);
        }

        Ok(Self { oid, nodes })
    }
}

impl Node {
    pub fn untyped_oid(&self) -> UntypedOid {
        match self {
            Self::File(FileNode { oid, .. }) => oid.into_untyped(),
            Self::Tree { oid, .. } => oid.into_untyped(),
        }
    }

    pub fn name(&self) -> &BStr {
        match self {
            Node::File(FileNode { name, .. }) | Node::Tree { name, .. } => name.as_bstr(),
        }
    }

    fn deserialize(mut data: impl BufRead) -> Result<Option<Self>, DeserializeError> {
        let mut mode = Vec::new();
        let bytes_read = data.read_until(b' ', &mut mode)?;
        if bytes_read == 0 {
            return Ok(None);
        }
        mode.pop().unwrap();

        let mut name = BString::from(Vec::new());
        data.read_until(b'\0', &mut name)?;
        name.pop().unwrap();

        let mut oid = [0; OID_SIZE];
        data.read_exact(&mut oid)?;
        let oid = UntypedOid::new(oid);

        let entry = if mode == Tree::MODE {
            Self::Tree {
                oid: oid.to_typed(),
                name,
            }
        } else {
            let mode = stat::Mode::from_base8(mode.as_bstr());
            Self::File(FileNode {
                oid: oid.to_typed(),
                mode,
                name,
            })
        };

        Ok(Some(entry))
    }
}

type BuilderNodes = BTreeMap<BString, SerializeNode>;

#[derive(Debug, Clone, Eq, PartialEq)]
pub struct Builder {
    trees: Vec<Option<BuilderNodes>>,
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub struct EntryBuilder {
    pub oid: Oid<Blob>,
    pub path: WsPath,
    pub mode: stat::Mode,
}

impl Builder {
    pub fn new() -> Self {
        Self {
            trees: vec![Some(BTreeMap::new())],
        }
    }

    pub fn entries(mut self, entries: impl IntoIterator<Item = EntryBuilder>) -> Self {
        for desc in entries {
            let mut parent = 0;

            for name in desc.path.parent_components() {
                let next = if let Some(existing) = self.tree(parent).get(name) {
                    match existing {
                        SerializeNode::Tree(existing) => *existing,
                        SerializeNode::Entry { .. } => panic!("Directory has same name as file"),
                    }
                } else {
                    self.trees.push(Some(BTreeMap::new()));
                    self.trees.len() - 1
                };

                self.tree(parent)
                    .insert(name.to_owned(), SerializeNode::Tree(next));

                parent = next;
            }

            self.tree(parent).insert(
                desc.path.file_name().to_owned(),
                SerializeNode::Entry {
                    oid: desc.oid,
                    mode: desc.mode,
                },
            );
        }

        self
    }

    fn tree(&mut self, i: usize) -> &mut BuilderNodes {
        self.trees[i].as_mut().expect("Subroot remains")
    }

    fn store_subroot(&mut self, db: &Db, subroot: usize) -> db::StoreResult<Tree> {
        let mut out = BString::from(Vec::new());

        for (name, entry) in self.trees[subroot].take().expect("Not already serialized") {
            let (oid, mode) = match entry {
                SerializeNode::Entry { oid, mode } => {
                    (oid.into_untyped(), mode.as_base8().as_bytes())
                }
                SerializeNode::Tree(tree) => {
                    let oid = self.store_subroot(db, tree)?;
                    (oid.into_untyped(), Tree::MODE)
                }
            };

            out.extend_from_slice(mode);
            out.push(b' ');
            out.extend_from_slice(name.as_bytes());
            out.push(b'\0');
            out.extend_from_slice(oid.as_bytes());
        }

        db.store_bytes::<Self>(&out)
    }
}

impl Default for Builder {
    fn default() -> Self {
        Self::new()
    }
}

impl ObjectBuilder for Builder {
    fn store(mut self, db: &Db) -> db::StoreResult<Tree> {
        self.store_subroot(db, 0)
    }

    type Object = Tree;
}

#[derive(Debug, Clone, Eq, PartialEq)]
enum SerializeNode {
    Tree(usize),
    Entry { oid: Oid<Blob>, mode: stat::Mode },
}

#[derive(Debug, displaydoc::Display, thiserror::Error)]
/// Error deserializing tree
pub struct DeserializeError(#[from] io::Error);

#[cfg(test)]
mod test {
    use insta::assert_debug_snapshot;

    use crate::core::stat::Mode;

    use super::*;

    #[test]
    fn builder_entries() {
        let builder = Builder::new().entries(vec![
            EntryBuilder {
                oid: Oid::zero(),
                path: WsPath::new_unchecked("top_level"),
                mode: Mode::Regular,
            },
            EntryBuilder {
                oid: Oid::zero(),
                path: WsPath::new_unchecked("top_level2"),
                mode: Mode::Regular,
            },
            EntryBuilder {
                oid: Oid::zero(),
                path: WsPath::new_unchecked("singly_nested/child"),
                mode: Mode::Regular,
            },
            EntryBuilder {
                oid: Oid::zero(),
                path: WsPath::new_unchecked("doubly_nested/inner/child"),
                mode: Mode::Regular,
            },
        ]);
        assert_debug_snapshot!(builder);
    }

    #[test]
    fn builder_entries_more_complex() {
        let builder = Builder::new().entries(vec![
            EntryBuilder {
                oid: Oid::zero(),
                path: WsPath::new_unchecked("f"),
                mode: Mode::Regular,
            },
            EntryBuilder {
                oid: Oid::zero(),
                path: WsPath::new_unchecked("dir_1/f"),
                mode: Mode::Regular,
            },
            EntryBuilder {
                oid: Oid::zero(),
                path: WsPath::new_unchecked("dir_2/dir_a/dir_x/f"),
                mode: Mode::Regular,
            },
            EntryBuilder {
                oid: Oid::zero(),
                path: WsPath::new_unchecked("dir_1/f2"),
                mode: Mode::Regular,
            },
            EntryBuilder {
                oid: Oid::zero(),
                path: WsPath::new_unchecked("dir_1/dir_a/f"),
                mode: Mode::Regular,
            },
            EntryBuilder {
                oid: Oid::zero(),
                path: WsPath::new_unchecked("dir_1/dir_a/dir_x/f"),
                mode: Mode::Regular,
            },
            EntryBuilder {
                oid: Oid::zero(),
                path: WsPath::new_unchecked("dir_1/dir_a/dir_y/f"),
                mode: Mode::Regular,
            },
        ]);
        assert_debug_snapshot!(builder);
    }
}
