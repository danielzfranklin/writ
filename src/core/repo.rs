use std::{
    collections::BTreeMap,
    env, fmt, fs,
    io::{self},
    path::{Path, PathBuf},
};

use crate::core::{
    db::{self, object, tree, Blob, Commit, Tree},
    index::{
        self,
        entry::{self, Entry, StatusChatty},
    },
    refs,
    ws::{self, ListFilesError, ReadFileError, StatFileError},
    Db, FileStatus, Index, IndexMut, ObjectBuilder, Oid, Refs, Status, Workspace, WsPath,
};
use chrono::Local;
use tracing::{debug, instrument};

#[derive(Debug, Clone)]
pub struct Repo {
    git_dir: PathBuf,
    pub workspace: Workspace,
    pub db: Db,
    pub refs: Refs,
    pub index: Index,
}

impl Repo {
    #[instrument(err)]
    pub fn new(workspace: impl Into<PathBuf> + fmt::Debug) -> Result<Self, ReadError> {
        let workspace_dir = workspace.into();
        let workspace_dir = workspace_dir
            .canonicalize()
            .map_err(|e| ReadError::Io(workspace_dir, e))?;

        let git_dir = workspace_dir.join(".git");

        if !git_dir
            .try_exists()
            .map_err(|e| ReadError::Io(git_dir.clone(), e))?
        {
            return Err(ReadError::NotRepo(workspace_dir));
        }

        let workspace = Workspace::new(workspace_dir);
        let db = Db::new(&git_dir);
        let refs = Refs::new(&git_dir);
        let index = Index::load(&git_dir)?;

        Ok(Self {
            git_dir,
            workspace,
            db,
            refs,
            index,
        })
    }

    pub fn for_current_dir() -> Result<Self, ForCurrentDirError> {
        let dir = env::current_dir()?;
        Ok(Self::new(dir)?)
    }

    #[instrument(err)]
    pub fn init(workspace: impl Into<PathBuf> + fmt::Debug) -> Result<Self, InitError> {
        let workspace_dir = workspace.into();

        fs::create_dir_all(&workspace_dir)
            .map_err(|e| InitError::CreateWorkspace(workspace_dir.clone(), e))?;

        let git_dir = workspace_dir.join(".git");
        if git_dir
            .try_exists()
            .map_err(|e| InitError::Open(git_dir.clone(), e))?
        {
            return Err(InitError::Exists(git_dir));
        }

        for child in &["objects", "refs"] {
            let child = git_dir.join(child);
            fs::create_dir_all(&child).map_err(|e| InitError::Write(child, e))?;
        }

        let workspace = Workspace::new(workspace_dir);
        let db = Db::new(&git_dir);
        let refs = Refs::new(&git_dir);
        let index = Index::load(&git_dir)?;

        Ok(Self {
            git_dir,
            workspace,
            db,
            refs,
            index,
        })
    }

    #[instrument(err)]
    pub fn add<I, P>(&mut self, files: I) -> Result<(), AddError>
    where
        I: IntoIterator<Item = P> + fmt::Debug,
        P: AsRef<Path>,
    {
        let workspace = &self.workspace;
        let db = &self.db;
        self.index.reload()?;
        let mut index = self.index.modify()?;

        for file in workspace.find_files(files)? {
            let data = workspace.read_file(&file)?;
            let stat = workspace.stat(&file)?;

            let oid = db::blob::Builder::new(data).store(&db)?;
            let entry = Entry::new(file, oid, stat);

            debug!("Adding {:?}", entry);
            index.add(entry);
        }

        index.commit()?;

        Ok(())
    }

    #[instrument(err)]
    pub fn commit(
        &mut self,
        name: impl Into<String> + fmt::Debug,
        email: impl Into<String> + fmt::Debug,
        msg: impl Into<String> + fmt::Debug,
    ) -> Result<(), CommitError> {
        let mut msg = msg.into();
        if msg.is_empty() {
            return Err(CommitError::EmptyMessage);
        }
        if !msg.ends_with('\n') {
            msg.push('\n');
        }

        let name = name.into();
        let email = email.into();

        let db = &self.db;
        let refs = &self.refs;
        let index = &self.index;

        let entries = index.entries().map(|entry| db::tree::EntryBuilder {
            oid: entry.oid,
            path: entry.path.clone(),
            mode: entry.mode(),
        });

        let root = db::tree::Builder::new().entries(entries).store(&db)?;

        let parent = refs.head()?;
        let author = db::Author::new_local(name, email, Local::now());
        let commit = db::commit::Builder::new(parent, root, author, msg).store(db)?;
        refs.update_head(&commit)?;

        Ok(())
    }

    /// Unlike git, this lists files only. Children of untracked directories are
    /// reported instead of reporting the directory itself.
    #[instrument(err)]
    pub fn status(&mut self) -> Result<BTreeMap<WsPath, FileStatus>, StatusError> {
        let head = if let Some(head) = self.refs.head()? {
            let tree = self.db.load(head)?.tree;
            self.db.load_tree_files(&WsPath::root(), tree)?
        } else {
            BTreeMap::new()
        };

        self.index.reload()?;
        let mut index = self
            .index
            .modify()
            .map_err(|e| StatusError::UpdateIndex(e.into()))?;

        let work = &self.workspace;

        let mut ws_statuses = BTreeMap::new();
        let mut index_statuses = BTreeMap::new();

        for path in self.workspace.list_files()? {
            let ws_status = Self::workspace_status_of(work, &mut index, &path)?;
            let index_status = Self::index_status_of(&index, &head, &path)?;
            debug!("{path} in workspace, so ws: {ws_status:?}, idx: {index_status:?}");
            ws_statuses.insert(path.clone(), ws_status);
            index_statuses.insert(path, index_status);
        }

        for entry in index.entries() {
            if !ws_statuses.contains_key(&entry.path) {
                debug!("{} in idx but not ws, so ws: Status::Deleted", entry.path);
                ws_statuses.insert(entry.path.clone(), Status::Deleted);
            }
        }

        for (path, _file) in head {
            if !index.is_tracked_file(&path) {
                debug!("{path} in head but not idx, so idx: Status::Deleted",);
                index_statuses.insert(path, Status::Deleted);
            }
        }

        index
            .commit()
            .map_err(|e| StatusError::UpdateIndex(e.into()))?;

        let mut statuses = BTreeMap::new();

        for (path, ws_status) in ws_statuses {
            let index_status = index_statuses.remove(&path).unwrap_or(Status::Untracked);
            statuses.insert(
                path.clone(),
                FileStatus {
                    path,
                    workspace: ws_status,
                    index: index_status,
                },
            );
        }

        for (path, index_status) in index_statuses {
            statuses.insert(
                path.clone(),
                FileStatus {
                    path,
                    workspace: Status::Deleted,
                    index: index_status,
                },
            );
        }

        Ok(statuses)
    }

    pub fn workspace_status_of(
        work: &Workspace,
        index: &mut IndexMut,
        path: &WsPath,
    ) -> Result<Status, StatusError> {
        let status = if let Some(entry) = index.entry(path) {
            match entry.index_status_chatty(work)? {
                StatusChatty::Unmodified => Status::Unmodified,
                StatusChatty::UnmodifiedButNewStat(new_stat) => {
                    index.update_stat(path, new_stat).expect("Entry exists");
                    Status::Unmodified
                }
                StatusChatty::Modified => Status::Modified,
                StatusChatty::Deleted => Status::Deleted,
            }
        } else {
            Status::Untracked
        };
        Ok(status)
    }

    #[allow(clippy::option_if_let_else)]
    pub fn index_status_of(
        index: &Index,
        head: &BTreeMap<WsPath, tree::FileNode>,
        path: &WsPath,
    ) -> Result<Status, StatusError> {
        let index_entry = if let Some(index_entry) = index.entry(path) {
            index_entry
        } else {
            return Ok(Status::Untracked);
        };

        let status = if let Some(head_file) = head.get(path) {
            if head_file.mode == index_entry.mode() && head_file.oid == index_entry.oid {
                Status::Unmodified
            } else {
                Status::Modified
            }
        } else {
            Status::Added
        };

        Ok(status)
    }

    pub fn show_head(&mut self) -> eyre::Result<()> {
        let head = self.refs.head()?.ok_or_else(|| eyre::eyre!("No HEAD"))?;
        let commit = self.db.load::<Commit>(head)?;
        eprintln!("HEAD: {}\n", head);
        self.print_tree(commit.tree, 0)?;
        Ok(())
    }

    fn print_tree(&mut self, tree: Oid<Tree>, level: usize) -> eyre::Result<()> {
        let tree = self.db.load::<Tree>(tree)?;
        let level_prefix = " ".repeat(level * 4);
        for node in tree.direct_children() {
            match node {
                db::tree::Node::File(db::tree::FileNode { name, mode, oid }) => {
                    println!("{}{} {} ({:?})", level_prefix, oid, name, mode)
                }
                db::tree::Node::Tree { name, oid } => {
                    println!("{}{} {}/", level_prefix, oid, name);
                    self.print_tree(*oid, level + 1)?;
                }
            }
        }
        Ok(())
    }
}

#[derive(Debug, displaydoc::Display, thiserror::Error)]
// Failed in initialize repository
pub enum InitError {
    /// Directory {0:?} already exists
    Exists(PathBuf),
    /// Failed to create workspace directory {0:?}
    CreateWorkspace(PathBuf, #[source] io::Error),
    /// Failed to open directory {0:?} to initialize
    Open(PathBuf, #[source] io::Error),
    /// Failed to populate {0:?}
    Write(PathBuf, #[source] io::Error),
    /// Failed to open index
    OpenIndex(#[from] index::LoadError),
}

#[derive(Debug, thiserror::Error, displaydoc::Display)]
/// Failed to read a directory as a git repository.
pub enum ReadError {
    /// The directory {0:?} is not a git repository
    NotRepo(PathBuf),
    /// IO error while checking if directory {0:?} is a git repository
    Io(PathBuf, #[source] io::Error),
    /// Failed to open index
    OpenIndex(#[from] index::LoadError),
}

#[derive(Debug, thiserror::Error, displaydoc::Display)]
/// Failed to read the current directory as a git repository.
pub enum ForCurrentDirError {
    /// Failed to get the current directory
    GetCurrentDir(#[from] io::Error),
    /// Failed to read git repository
    ReadError(#[from] ReadError),
}

#[derive(Debug, thiserror::Error, displaydoc::Display)]
pub enum AddError {
    /// Failed to reload index
    ReloadIndex(#[from] index::LoadError),
    /// Failed to open index of modifications
    OpenIndex(#[from] index::OpenForModificationsError),
    /// Failed to find files provided in repository
    FindFiles(#[from] ws::ListFilesError),
    /// Failed to stat file
    Stat(#[from] StatFileError),
    /// Failed to read file
    Read(#[from] ReadFileError),
    /// Failed to store file
    StoreBlob(#[from] db::StoreError<Blob>),
    /// Failed to commit changes to index
    CommitIndex(#[from] index::CommitError),
}

#[derive(Debug, thiserror::Error, displaydoc::Display)]
pub enum CommitError {
    /// Empty commit message
    EmptyMessage,
    /// Failed to load index
    LoadIndex(#[from] index::OpenForModificationsError),
    /// Failed to store tree
    StoreTree(#[from] db::StoreError<Tree>),
    /// Failed to store commit
    StoreCommit(#[from] db::StoreError<Commit>),
    /// Failed to read ref
    ReadRef(#[from] refs::ReadError),
    /// Failed to parse oid of parent commit
    ParseParentOid(#[from] object::ParseOidError),
    /// Failed to update ref
    UpdateRef(#[from] refs::UpdateError),
}

#[derive(Debug, thiserror::Error, displaydoc::Display)]
pub enum StatusError {
    /// Failed to reload index
    ReloadIndex(#[from] index::LoadError),
    /// Failed to get head oid
    GetHeadOid(#[from] refs::ReadError),
    /// Failed to load head commit
    LoadHeadCommit(#[from] db::LoadError<db::Commit>),
    /// Failed to load of tree from head
    LoadHeadTree(#[from] db::LoadError<db::Tree>),
    /// Failed to list files
    ListFiles(#[from] ListFilesError),
    /// Failed to check if file unchanged
    IsUnchanged(#[from] entry::IsUnchangedError),
    /// Failed to update index with new stat
    UpdateIndex(#[from] index::ModifyError),
}
