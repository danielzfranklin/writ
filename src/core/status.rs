use crate::core::WsPath;

#[derive(Debug, Clone, Eq, PartialEq)]
#[allow(clippy::module_name_repetitions)]
pub struct FileStatus {
    pub path: WsPath,
    pub index: Status,
    pub workspace: Status,
}

#[derive(Debug, Clone, Copy, Eq, PartialEq)]
pub enum Status {
    Untracked,
    Modified,
    Unmodified,
    Deleted,
    Added,
}

impl Status {
    pub fn name(self) -> &'static str {
        match self {
            Status::Untracked => "untracked",
            Status::Modified => "modified",
            Status::Unmodified => "unmodified",
            Status::Deleted => "deleted",
            Status::Added => "added",
        }
    }
}
