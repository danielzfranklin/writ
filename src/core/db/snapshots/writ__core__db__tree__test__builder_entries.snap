---
source: src/core/db/tree.rs
expression: builder

---
Builder {
    trees: [
        Some(
            {
                "doubly_nested": Tree(
                    2,
                ),
                "singly_nested": Tree(
                    1,
                ),
                "top_level": Entry {
                    oid: Oid {
                        bytes: "0000000000000000000000000000000000000000",
                        type: "blob",
                    },
                    mode: Regular,
                },
                "top_level2": Entry {
                    oid: Oid {
                        bytes: "0000000000000000000000000000000000000000",
                        type: "blob",
                    },
                    mode: Regular,
                },
            },
        ),
        Some(
            {
                "child": Entry {
                    oid: Oid {
                        bytes: "0000000000000000000000000000000000000000",
                        type: "blob",
                    },
                    mode: Regular,
                },
            },
        ),
        Some(
            {
                "inner": Tree(
                    3,
                ),
            },
        ),
        Some(
            {
                "child": Entry {
                    oid: Oid {
                        bytes: "0000000000000000000000000000000000000000",
                        type: "blob",
                    },
                    mode: Regular,
                },
            },
        ),
    ],
}
